### PR TITLE
feat: implement gRPC server 

### PR DESCRIPTION
**Pull request by aramase**
**What this PR does / why we need it**:
- Adds proto definitions and implements the GRPC provider server 

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by aramase** [comment]
This PR is ready for review. I've added my custom branch currently to run e2e tests as there is currently no driver release with the GRPC changes.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -89,43 +90,44 @@ jobs:
...
-  - script: |
-      sudo add-apt-repository ppa:rmescandon/yq && sudo apt-get -o Acquire::Retries=30 update && sudo apt-get install yq
-      sudo apt-get -o Acquire::Retries=30 update && sudo apt-get -o Acquire::Retries=30 install -y bats
-      bats --version
-    displayName: 'Set up workspace and install dependencies'
-  - script: |
-      # Generate image version
-      IMAGE_VERSION=linux-e2e-$(git rev-parse --short HEAD)
-      echo "Image version: ${IMAGE_VERSION}"
+  # steps:
```
Comment (ritazh) [inline azure-pipelines.yml:93]: why is this all commented out?

File: cmd/main.go [hunk 2]
```diff
@@ -23,106 +20,27 @@ var (
 	debug          = pflag.Bool("debug", false, "sets log to debug level")
 	versionInfo    = pflag.Bool("version", false, "prints the version information")
 	contextTimeout = pflag.Int("context-timeout", 110, "context timeout in seconds for provider calls")
+	endpoint       = pflag.String("endpoint", "unix://tmp/csi.sock", "CSI endpoint")
```
Comment (ritazh) [inline cmd/main.go:23]: nit: should we make the default name more descriptive e.g. `azure.sock` and description `CSI grpc endpoint`

File: cmd/main.go [hunk 3]
```diff
@@ -23,106 +20,27 @@ var (
 	debug          = pflag.Bool("debug", false, "sets log to debug level")
```
Comment (ritazh) [inline cmd/main.go:20]: where is this flag used now or how do users set the log levels now?

File: deployment/provider-azure-installer-windows.yaml [hunk 4]
```diff
@@ -21,27 +21,39 @@ spec:
         app: csi-secrets-store-provider-azure
     spec:
       serviceAccountName: csi-secrets-store-provider-azure
+      hostNetwork: true
```
Comment (ritazh) [inline deployment/provider-azure-installer-windows.yaml:24]: For these yaml changes, we should wait until after this https://github.com/Azure/secrets-store-csi-driver-provider-azure/pull/215 then move these to manifest_staging

File: deployment/provider-azure-installer-windows.yaml [hunk 5]
```diff
@@ -21,27 +21,39 @@ spec:
...
             - name: TARGET_DIR
-              value: "C:\\k\\secrets-store-csi-providers"
+              value: "C:\\provider"
           volumeMounts:
-            - mountPath: "C:\\k\\secrets-store-csi-providers"
+            - mountPath: "C:\\provider"
               name: providervol
+            - name: mountpoint-dir
+              mountPath: "C:\\var\\lib\\kubelet\\pods"
+              mountPropagation: Bidirectional
```
Comment (ritazh) [inline deployment/provider-azure-installer-windows.yaml:46]: why do we need bidirectional now?

File: pkg/provider/provider.go [hunk 6]
```diff
@@ -163,7 +129,7 @@ func (p *Provider) initializeKvClient() (*kv.BaseClient, error) {
 	if err != nil {
 		return nil, errors.Wrapf(err, "failed to add user agent to keyvault client")
 	}
-	token, err := p.GetKeyvaultToken(AuthGrantType())
+	token, err := p.GetKeyvaultToken()
```
Comment (ritazh) [inline pkg/provider/provider.go:132]: 👍  Thanks for cleaning this up.

File: pkg/provider/provider.go [hunk 7]
```diff
@@ -669,3 +514,16 @@ func validateObjectFormat(objectFormat, objectType string) error {
 	}
 	return nil
 }
+
+// getObjectVersion parses the id to retrieve the version
+// of object fetched
+func getObjectVersion(id string) string {
+	splitID := strings.Split(id, "/")
+	return splitID[len(splitID)-1]
```
Comment (ritazh) [inline pkg/provider/provider.go:522]: Is this logic always true? Where is this documented?

Comment (aramase) [reply pkg/provider/provider.go:522]: Opened an issue - https://github.com/Azure/azure-rest-api-specs/issues/10825

File: deployment/provider-azure-installer-windows.yaml (hunk 5 above)
Comment (aramase) [reply deployment/provider-azure-installer-windows.yaml]: https://kubernetes-csi.github.io/docs/deploying.html#driver-volume-mounts

```
The mount point used by the CSI driver must be set to Bidirectional to allow Kubelet on the host to see mounts created by the CSI driver container. See the example below:
```

If we don't set it to `Bidirectional` then the contents written to the target path will not show up in the pod mount.

File: .pipelines/templates/aks-setup.yaml [hunk 8]
```diff
@@ -13,7 +13,6 @@ steps:
         -n ${AZURE_CLUSTER_NAME} \
         --subscription ${SUBSCRIPTION_ID} \
         --node-count ${AGENT_COUNT} \
-        --kubernetes-version ${AKS_KUBERNETES_VERSION} \
```
Comment (aramase) [inline .pipelines/templates/aks-setup.yaml:16]: Removing this so we can run the latest default for AKS. Without this we would need to update the `AKS_KUBERNETES_VERSION` if the current version is out of support in AKS.

**Comment by aramase** [comment]
@ritazh This PR is ready for review.

File: azure-pipelines.yml [hunk 9]
```diff
@@ -125,7 +125,7 @@ jobs:
       TENANT_ID: $(TENANT_ID)
       CI_KIND_CLUSTER: true
       AZURE_ENVIRONMENT: AzureStackCloud
-      AZURE_ENVIRONMENT_FILEPATH: /etc/kubernetes/secrets-store-csi-providers/custom_environment.json
+      AZURE_ENVIRONMENT_FILEPATH: /provider/custom_environment.json
```
Comment (ritazh) [inline azure-pipelines.yml:128]: what populates `/provider/custom_environment.json`? 
I see it was created by https://github.com/Azure/secrets-store-csi-driver-provider-azure/blob/master/test/kind-config.yaml#L7

File: cmd/main.go [hunk 10]
```diff
@@ -1,128 +1,42 @@
...
 )
 
 var (
-	attributes     = pflag.String("attributes", "", "volume attributes")
-	secrets        = pflag.String("secrets", "", "node publish ref secret")
-	targetPath     = pflag.String("targetPath", "", "Target path to write data.")
-	permission     = pflag.String("permission", "", "File permission")
-	debug          = pflag.Bool("debug", false, "sets log to debug level")
-	versionInfo    = pflag.Bool("version", false, "prints the version information")
-	contextTimeout = pflag.Int("context-timeout", 110, "context timeout in seconds for provider calls")
```
Comment (ritazh) [inline cmd/main.go:25]: How do we treat context timeout here for grpc server and client in the driver?

Comment (aramase) [reply cmd/main.go:25]: The client in CSI driver passes the kubelet context to the gRPC server as part of the RPC call (https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/pkg/secrets-store/provider_client.go#L104) . The server passes the context to all the underlying functions.

File: azure-pipelines.yml (hunk 9 above)
Comment (aramase) [reply azure-pipelines.yml:128]: `/provider` is the container path for `/etc/kubernetes/secrets-store-csi-driver-providers` host path. As the provider mounts that volume, `custom_environment.json` is available in the pod.

File: pkg/provider/provider.go [hunk 11]
```diff
@@ -218,278 +158,184 @@ func (p *Provider) getVaultURL(ctx context.Context) (vaultURL *string, err error
...
 	if keyvaultName == "" {
-		return fmt.Errorf("keyvaultName is not set")
+		return nil, fmt.Errorf("keyvaultName is not set")
 	}
 	if tenantID == "" {
-		return fmt.Errorf("tenantId is not set")
+		return nil, fmt.Errorf("tenantId is not set")
+	}
+	if len(usePodIdentityStr) == 0 {
+		usePodIdentityStr = "false"
```
Comment (ritazh) [inline pkg/provider/provider.go:183]: if `usePodIdentityStr` is not used anywhere else, we should set `usePodIdentity` here

File: pkg/provider/provider.go [hunk 12]
```diff
@@ -218,278 +158,184 @@ func (p *Provider) getVaultURL(ctx context.Context) (vaultURL *string, err error
...
+	}
+	if len(useVMManagedIdentityStr) == 0 {
+		useVMManagedIdentityStr = "false"
```
Comment (ritazh) [inline pkg/provider/provider.go:186]: same here

File: pkg/provider/provider.go [hunk 13]
```diff
@@ -218,278 +158,184 @@ func (p *Provider) getVaultURL(ctx context.Context) (vaultURL *string, err error
...
 	p.AzureCloudEnvironment = azureCloudEnv
-	p.UsePodIdentity = usePodIdentity
-	p.UseVMManagedIdentity = useVMManagedIdentity
-	p.UserAssignedIdentityID = userAssignedIdentityID
 	p.TenantID = tenantID
 
+	objectVersionMap := make(map[string]string)
 	for _, keyVaultObject := range keyVaultObjects {
-		log.Infof("fetching object: %s, type: %s from key vault", keyVaultObject.ObjectName, keyVaultObject.ObjectType)
+		klog.Infof("fetching object: %s, type: %s from key vault %s", keyVaultObject.ObjectName, keyVaultObject.ObjectType, p.KeyvaultName)
```
Comment (ritazh) [inline pkg/provider/provider.go:251]: should this be `klog.V(2).Infof` to be consistent with when we want to log these values?

File: pkg/utils/grpc.go [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+
+	klog.Infof("Listening for connections on address: %v", listener.Addr())
+
+	err = server.Serve(listener)
+	if err != nil {
+		klog.Fatalf("Failed to serve: %v", err)
+	}
+}
+
+func ParseEndpoint(ep string) (string, string, error) {
```
Comment (ritazh) [inline pkg/utils/grpc.go:94]: we need unit test for this file

File: pkg/provider/server.go [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/Azure/secrets-store-csi-driver-provider-azure/pkg/utils"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"sigs.k8s.io/secrets-store-csi-driver/provider/v1alpha1"
+)
+
+// CSIDriverProviderServer providers a Secrets Store CSI Driver provider implementation
+type CSIDriverProviderServer struct {
```
Comment (ritazh) [inline pkg/provider/server.go:19]: we need unit test for this file

File: azure-pipelines.yml (hunk 9 above)
Comment (ritazh) [reply azure-pipelines.yml:128]: Ok thanks! I see you updated the mount path: `mountPath: "/provider"`

File: cmd/main.go (hunk 10 above)
Comment (ritazh) [reply cmd/main.go:25]: do we want to set timeouts for client and server?

File: pkg/utils/grpc.go (hunk 14 above)
Comment (aramase) [reply pkg/utils/grpc.go:14]: Refactored this a little bit more and added unit tests for the helpers

File: pkg/provider/provider.go (hunk 12 above)
Comment (aramase) [reply pkg/provider/provider.go:186]: This is being used - https://github.com/Azure/secrets-store-csi-driver-provider-azure/pull/203/files#diff-6155820fac2dd2fda7353505b1f69319R198-R209

File: pkg/provider/provider.go (hunk 11 above)
Comment (aramase) [reply pkg/provider/provider.go:183]: This is being used - https://github.com/Azure/secrets-store-csi-driver-provider-azure/pull/203/files#diff-6155820fac2dd2fda7353505b1f69319R198-R209

File: cmd/main.go (hunk 10 above)
Comment (aramase) [reply cmd/main.go:25]: The parent context is the kubelet context which has a timeout of 2m3s. The same context is propagated across as part of the RPC call. We pass the context to all Azure API calls as well. So the parent context timeout/deadline is essentially the value we want to honor in the provider. The upper ceiling for the timeout is how long kubelet can wait for the driver-provider communication to complete. I don't see a scenario where we would want to override that context with a shorter time. WDYT?

Comment (ritazh) [reply cmd/main.go:25]: okay that makes sense!

File: pkg/provider/provider.go (hunk 11 above)
Comment (ritazh) [reply pkg/provider/provider.go:183]: This is a nit so feel free to ignore it. 

I meant why not set `usePodIdentity` immediate following this? ie. move up https://github.com/Azure/secrets-store-csi-driver-provider-azure/blob/66c131910d8da0e9aadae02d7260fd30fb245077/pkg/provider/provider.go#L198-L210

File: cmd/main.go [hunk 16]
```diff
@@ -1,27 +1,39 @@
...
-	endpoint    = pflag.String("endpoint", "unix://tmp/azure.sock", "CSI gRPC endpoint")
+	endpoint    = pflag.String("endpoint", "unix:///tmp/azure.sock", "CSI gRPC endpoint")
 )
 
 func main() {
 	klog.InitFlags(nil)
 	pflag.Parse()
 
+	signalChan := make(chan os.Signal, 1)
+	signal.Notify(signalChan, syscall.SIGTERM, syscall.SIGINT, os.Interrupt)
```
Comment (ritazh) [inline cmd/main.go:35]: +1

File: pkg/provider/provider.go (hunk 11 above)
Comment (aramase) [reply pkg/provider/provider.go:183]: Yeah, that makes sense! Moved it up.

**Review by ritazh — APPROVED** [review]

**Pull request merged**